testing/mock_raster_cache.h

```diff
@@ -0,0 +1,92 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef TESTING_MOCK_RASTER_CACHE_H_
+#define TESTING_MOCK_RASTER_CACHE_H_
+
+#include "flutter/flow/raster_cache.h"
+#include "third_party/skia/include/core/SkImage.h"
+#include "third_party/skia/include/core/SkPicture.h"
+
+namespace flutter {
+namespace testing {
+
+// Mock |RasterCache|, useful for writing tests that impact the Flutter engine
+// raster_cache without requiring a GPU.
+//
+// The |MockCanvas| stores a list of Engine Layers and Pictures that have been
+// inserted into the cache that the test can later verify against the expected
+// list of cached objects.
+class MockRasterCache : public RasterCache {
+ public:
+  // Return true if the cache is generated.
+  //
+  // We may return false and not generate the cache if
+  // 1. The picture is not worth rasterizing
+  // 2. The matrix is singular
+  // 3. The picture is accessed too few times
+  // 4. There are too many pictures to be cached in the current frame.
+  //    (See also kDefaultPictureCacheLimitPerFrame.)
+  bool Prepare(GrContext* context,
+               SkPicture* picture,
+               const SkMatrix& transformation_matrix,
+               SkColorSpace* dst_color_space,
+               bool is_complex,
+               bool will_change) override;
+
+  void Prepare(PrerollContext* context,
+               Layer* layer,
+               const SkMatrix& ctm) override;
+
+  bool Draw(const SkPicture& picture, SkCanvas& canvas) const override;
+
+  bool Draw(const Layer* layer,
+            SkCanvas& canvas,
+            SkPaint* paint = nullptr) const override;
+
+  void SweepAfterFrame();
+
+  void Clear();
+
+  int PictureCacheCount() { return picture_cache_.size(); }
+  int LayerCacheCount() { return layer_cache_.size(); }
+
+  void SetCheckboardCacheImages(bool checkerboard);
+
+  size_t GetCachedEntriesCount() const;
+
+  MockRasterCache();
+
+ private:
+  struct MockEntry {
```

Comment:
We might need to remove `RasterCache::Entry` and add its fields directly into `RasterCacheResult` so you can just use `MockRasterCacheResult` without having to create another `MockEntry`.

Reply:
Since the logic involving the other fields was external to the rasterize methods that I refactored, I left this as a separate structure.